base/meta.jl

```diff
@@ -223,6 +223,25 @@ function parse(str::AbstractString; raise::Bool=true, depwarn::Bool=true)
     return ex
 end
 
+"""
+    parsefile(filename)
+
+Parse the Julia source file located at `filename` and return a `:toplevel`
+expression that contains all expressions in the file.
+
```

Comment:
What happens when there's a parse error?

I think we should consider making this (a) consistent with `Meta.parse` and (b) documented.

Reply:
```julia-repl
shell> cat test.jl
a = 1
b = 2.2.2
c = 3

julia> parsefile("test.jl")
:($(Expr(:error, "invalid numeric constant \"2.2.\"")))
```

Reply:
I have added to the docstring explaining the behavior when there is a parse error, and I have added some tests.

Reply:
I'm not sure this is the right behavior. It doesn't make it possible to emulate `include`, which evaluates the expressions before the error. I guess we decided not to do that in the REPL, but I'm not sure why; maybe a REPL input is more of an atomic unit than a file. Anyway it seems to me that should be handled only by the REPL and not a parsing function.

Reply:
Agreed. I need to sort this out while refactoring the internals so that `include` goes via the same code path that we have here. For that it seems we need to return the list of correctly parsed statements, and also include the error somehow (at the end of the list? separately?)

For this user-facing API I'd favor emulating `Meta.parse` and the `raise` keyword. In fact I'm pretty sure everything should go via `Meta.parse`; we've just got to generalize `greedy` somehow and allow a filename to be passed.

For `greedy`, I'm thinking we should allow specification of grammar rules via `rule=:atom` vs `rule=:statement` vs `rule=:toplevel`? But I'm not sure this is a good general way to enter the current parser and whether it generalizes between parsers. Also we don't have a formal grammar...

Reply:
> we've just got to generalize `greedy` somehow

As you point out, we only seem to need 3 options. I don't have a strong opinion on what the interface should be though; a keyword argument with a few symbol values is not necessarily better than having 2-3 functions.

> Also we don't have a formal grammar...

If you really need to talk about specific grammar productions by name, each one is defined by a function called `parse-X`, so there you go. But, I really don't think it's too useful to make the exact set of productions part of the interface to `parse`. Even grammars written out in BNF will often have rules that are really just implementation details.